LockExample/FileService.cs

```diff
@@ -1,52 +1,50 @@
 ﻿using System.Diagnostics;
+using LockExample.Utils;
 
 namespace LockExample
 {
     internal static class FileService
     {
         private const string FolderName = "LockStatementApp";
         private const string Destination = "Destination";
-        private static readonly object fileLock = new();
-
-        public static void StartFileCopy()
+        public static async Task StartFileCopy()
         {
             var dirApp = Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData);
             var destinationPath = Path.Combine(dirApp, Destination);
             var sourcePath = Path.Combine(dirApp, FolderName);
-            if (!Directory.Exists(destinationPath))
-            {
-                _ = Directory.CreateDirectory(destinationPath);
-            }
-            _ = CopyFileAsync(sourcePath, destinationPath);
+            DirectoryUtil.EnsureDirectory(destinationPath);
+            await CopyFileAsync(sourcePath, destinationPath).ConfigureAwait(false);
         }
         public static async Task CopyFileAsync(string source, string destination)
         {
-            var stopWatch = new Stopwatch();
             var fileEntries = Directory.GetFiles(source);
-            var len = fileEntries.Length;
-            var tasks = new Task[len];
-            for (var i = 0; i < len; i++)
+            var tasks = new List<Task<ResultInfo>>();
+            foreach (var item in fileEntries)
             {
-                lock (fileLock)
-                {
-                    stopWatch.Start();
-                    Console.WriteLine(Path.GetFileName(fileEntries[i]));
-                    tasks[i] = Task.Run(() => CopyToDestination(source, destination, fileEntries[i]));
-                    stopWatch.Stop();
-                    TimeSpan ts = stopWatch.Elapsed;
-                    Console.WriteLine($"RunTime {ts.TotalSeconds * 1000}");
-                }
-                Thread.Sleep(2000);
-
+                var task = Task.Run(() => CopyToDestination(source, destination, item));
+                tasks.Add(task);
+            }
+            _ = await Task.WhenAll(tasks).ConfigureAwait(true);
+            foreach (Task<ResultInfo> item in tasks)
+            {
+                ResultInfo resultItem = item.GetAwaiter().GetResult();
```

Comment:
please try to refactor to have it written to file right after the file is copied, in `CopyToDestination` method